### ISSUE TITLE
How to design configuration object?

### ISSUE THREAD

**Issue by nohwnd** [pester/Pester#1412]
Pester will have a rather complex configuration object and it would be nice to immediately see what the defaults are, what the current value is and what it means. My goal is to give the user some context while keeping the object simple to use once he gets familiar with it.

### Navigating the object
Say I have an object with this structure: 

```powershell
$c = @{ 
    Should = @{ <- # Should configuration.
        ErrorAction = 'Stop' # <- "Controls if Should throws on error."
    }
}
```
Navigating through the object I don't see the descriptions, just values, so instead of this: 
```
$c.Should.ErrorAction 
# output
Stop
```

I would like to see: 

```
$c.Should.ErrorAction 
# output
Value    Default Description
-----    ------- -----------
Continue Stop    Controls if Should throws on error. Use 'Stop' to throw on error, or 'Continue' to fail at the end of the test.
```

And similarly for `$c.Should` and other sections of the config. 

### Setting values

User should not be forced to use special syntax like this: 
```powershell 
$c.Should.ErrorAction.Value = "Continue"
```
Instead just this should be possible: 

```powershell
$c.Should.ErrorAction = "Continue"
```

### Conclusion

I can achieve most of this via C# (and some xml templates later to get the ordering of properties correct), but am I reinventing the wheel here? Is there a simple powershell way to do this? 

```powershell


Add-Type -typeDefinition @'
public class Option<T>
{
    public Option(string description, T defaultValue, T value)
    {
        Default = defaultValue;
        Value = value;
        Description = description;
    }

    public T Default { get; private set; }
    public string Description { get; private set; }
    public T Value { get; set; }

    public static implicit operator Option<T>(T value)
    {
        return new Option<T>(string.Empty, value, value);
    }

    public static implicit operator T(Option<T> option)
    {
        return option.Value;
    }

    public override string ToString()
    {
        return string.Format("{0} ({1}, default: {2})", Description, Value, Default);
    }
}

public class StringOption: Option<string>
{
    public StringOption(string description, string defaultValue) : base(description, defaultValue, defaultValue)
    {

    }

    public StringOption(string description, string defaultValue, string value) : base(description, defaultValue, value) 
    {

    }

    public static implicit operator StringOption(string value)
    {
        return new StringOption(string.Empty, value, value);
    }

    public static implicit operator string(StringOption option)
    {
        return option.Value;
    }
}



public class PesterConfiguration
{
    public PesterConfiguration()
    {
        Should = new ShouldConfiguration();
    }
    public ShouldConfiguration Should { get; private set; }
}

public class ShouldConfiguration
{
    private StringOption _errorAction;

    public ShouldConfiguration()
    {
        ErrorAction = new StringOption("Controls if Should throws on error. Use 'Stop' to throw on error, or 'Continue' to fail at the end of the test.", "Stop");
    }

    public StringOption ErrorAction
    {
        get { return _errorAction; }
        set
        {
            if (_errorAction == null)
            {
                _errorAction = value;
            }
            else
            {
                _errorAction = new StringOption(_errorAction.Description, _errorAction.Default, value.Value);
            }
        }
    }
}
'@ -ErrorAction Stop

$c = [PesterConfiguration]::new()

$c.Should.ErrorAction = "Continue"

$c.Should.ErrorAction | Select-Object -property Value, Default, Description
```

**Comment by nohwnd** [comment]
Based on the feedback on slack I am sticking with this solution, and adopting creating the type from constructor that accepts a hashtable (idictionary). 

> Fred 2:16 PM
> Looks pretty nifty to me, given your declared goal
> Can't see how this could be done in a reasonably different manner, especially when considering v2 compat.
> You may want to include a "Name" and "Category" property in the option-object, in order to make it more visible what the option-object belongs to.
> Also: Some validation would probably nice to have for each option, to reduce error rate
> Also: If you give [PesterConfiguration] a constructor that accepts a hashtable, then parses that out, that would make declaring the whole thing a lot more convenient.

```powershell
$config = [PesterConfiguration]@{ 
    Should = @{ <- # Should configuration.
        ErrorAction = 'Stop' # <- "Controls if Should throws on error."
    }
}
```

**Comment by nohwnd** [comment]
Fixed in #1418
